dll/shellext/mydocs/CMyDocsDropHandler.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * PROJECT:     mydocs
+ * LICENSE:     LGPL-2.1+ (https://spdx.org/licenses/LGPL-2.1+)
+ * PURPOSE:     MyDocs implementation
+ * COPYRIGHT:   Copyright 2020 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
+ */
+
+#include "precomp.hpp"
+
+WINE_DEFAULT_DEBUG_CHANNEL(mydocs);
+
+CMyDocsDropHandler::CMyDocsDropHandler()
+{
+    InterlockedIncrement(&g_ModuleRefCnt);
+}
+
+CMyDocsDropHandler::~CMyDocsDropHandler()
+{
+    InterlockedDecrement(&g_ModuleRefCnt);
+}
+
+// IDropTarget
+STDMETHODIMP
+CMyDocsDropHandler::DragEnter(IDataObject *pDataObject, DWORD dwKeyState,
+                                POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    *pdwEffect &= DROPEFFECT_LINK;
+
+    return S_OK;
+}
+
+STDMETHODIMP
+CMyDocsDropHandler::DragOver(DWORD dwKeyState, POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    *pdwEffect &= DROPEFFECT_LINK;
+
+    return S_OK;
+}
+
+STDMETHODIMP CMyDocsDropHandler::DragLeave()
+{
+    TRACE("(%p)\n", this);
+    return S_OK;
+}
+
+STDMETHODIMP
+CMyDocsDropHandler::Drop(IDataObject *pDataObject, DWORD dwKeyState,
+                         POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    if (!pDataObject)
+    {
+        ERR("pDataObject is NULL\n");
+        *pdwEffect = 0;
+        DragLeave();
+        return E_POINTER;
+    }
+
+    STGMEDIUM stg;
+    FORMATETC etc = { CF_HDROP, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+    HRESULT hr = pDataObject->GetData(&etc, &stg);
+    if (FAILED_UNEXPECTEDLY(hr))
+    {
+        *pdwEffect = 0;
+        DragLeave();
+        return E_FAIL;
+    }
+
+    // "My Documents"
+    WCHAR szDir[MAX_PATH + 1];
+    SHGetSpecialFolderPathW(NULL, szDir, CSIDL_PERSONAL, FALSE);
+    szDir[lstrlenW(szDir) + 1] = 0;
+
+    HDROP hDrop = reinterpret_cast<HDROP>(stg.hGlobal);
+    UINT cItems = ::DragQueryFileW(hDrop, 0xFFFFFFFF, NULL, 0);
```

Comment:
I am not a big fan of extracting info this way, but I am not sure we have enough to do it better :(

Reply:
It seems like our `IDataObject` `HDROP` doesn't support multiple file items...

Reply:
Yeah, we should probably write a helper for this, that tries a few methods to extract files.
First using HIDA, then falling back to legacy systems like HDROP...

Reply:
I used HIDA. Done.

Reply:
I used HIDA. Done.